File: src/lib/utility-calc.js
/**
 * Converts True Airspeed (TAS) to Calibrated Airspeed (CAS).
 *
 * @param {number} tasKt - True Airspeed in knots
 * @param {number} pressAltFt - Pressure Altitude in feet (valid below ~36,089 ft / troposphere only)
 * @param {number} oatC - Outside Air Temperature in Celsius
 * @returns {number} Calibrated Airspeed (CAS) in knots
 */
export function convertTasToCas(tasKt, pressAltFt, oatC) {
    const P0 = 29.92126;
    const T0 = 288.15;
    const TAS_TO_FPS = 1.68781;
    const a0 = 1116.45;

    const T = oatC + 273.15;
    const delta = Math.pow(1 - (6.8755856e-6 * pressAltFt), 5.25588);
    const P = P0 * delta;
    const a = a0 * Math.sqrt(T / T0);

    const tasFps = tasKt * TAS_TO_FPS;
    const M = tasFps / a;

    let qc;
    if (M === 0) {
        qc = 0;
    } else {
        qc = P * (Math.pow(1 + 0.2 * M * M, 3.5) - 1);
    }

    let casKt;
    if (qc <= 0) {
        casKt = 0;
    } else {
        const casFps = a0 * Math.sqrt(5 * (Math.pow(qc / P0 + 1, 2 / 7) - 1));
        casKt = casFps / TAS_TO_FPS;
    }

    return casKt;
}

File: ios/AvCalcEngineKit/Sources/AvCalcEngineKit/UtilityCalc.swift
// Ported 1:1 from src/lib/utility-calc.js.

import Foundation

public enum UtilityCalc {
    /// Converts True Airspeed (TAS) to Calibrated Airspeed (CAS).
    /// - Parameters:
    ///   - tasKt: True Airspeed in knots
    ///   - pressAltFt: Pressure Altitude in feet (valid below ~36,089 ft / troposphere only)
    ///   - oatC: Outside Air Temperature in Celsius
    /// - Returns: Calibrated Airspeed (CAS) in knots
    public static func convertTasToCas(_ tasKt: Double, _ pressAltFt: Double, _ oatC: Double) -> Double {
        let P0 = 29.92126
        let T0 = 288.15
        let TAS_TO_FPS = 1.68781
        let a0 = 1116.45

        let T = oatC + 273.15
        let delta = pow(1 - (6.8755856e-6 * pressAltFt), 5.25588)
        let P = P0 * delta
        let a = a0 * sqrt(T / T0)

        let tasFps = tasKt * TAS_TO_FPS
        let M = tasFps / a

        let qc: Double
        if M == 0 {
            qc = 0
        } else {
            qc = P * (pow(1 + 0.2 * M * M, 3.5) - 1)
        }

        let casKt: Double
        if qc <= 0 {
            casKt = 0
        } else {
            let casFps = a0 * sqrt(5 * (pow(qc / P0 + 1, 2.0 / 7.0) - 1))
            casKt = casFps / TAS_TO_FPS
        }

        return casKt
    }
}
